### PR TITLE
client-go testing: allow to use generateName

### PR DESCRIPTION
**Pull request by aojea**
/kind bug
/kind failing-test

```release-note
NONE
```

Unit testing controllers with the fake client go, when using objects that only have GenerateName (and no name), can not obtain the Name of the object and use an empty string.

This PR implements the GenerateName logic used by the apiserver, so mocks are more realistic

### CODE REVIEW

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 1]
```diff
@@ -391,6 +392,20 @@ func (t *tracker) add(gvr schema.GroupVersionResource, obj runtime.Object, ns st
 		return err
 	}
 
+	// Only generate the name for objects that are Created
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/testing/fixture.go:395]: hmm... I'm not really a fan of replicating server logic/mutations in client-go in inaccurate ways (name generation is resource-specific)... are we also doing things like populating uid and creationTimestamp on creation?

Comment (aojea) [reply staging/src/k8s.io/client-go/testing/fixture.go:395]: > are we also doing things like populating uid and creationTimestamp on creation?

nope,  that parted surprised me too

> name generation is resource-specific

ah, right, this depends on the strategy

**Comment by aojea** [comment]
/close

https://github.com/kubernetes/kubernetes/pull/112280#discussion_r964939394

Thanks Jordan

**Pull request closed without merging**